### ISSUE TITLE
Wireguard: Routes Not Added for Allowed IPs Outside of Tunnel Address Subnet on 24.1.2 and Above

### ISSUE THREAD

**Issue by rfrederick** [opnsense/core#7304]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [x] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

**Describe the bug**

In 24.1.2 and above, Wireguard does not add routes for IP addresses contained in peers' Allowed IPs that are outside the subnet(s) of the associated instances' Tunnel Address(es).

**To Reproduce**

1. Create a Wireguard instance with defined IPv4 and IPv6 tunnel addresses, e.g. 172.16.40.1/24 fe80::2001:db8:1:1/64
2. Create one or more peers with Allowed IP entries containing addresses within the instance's tunnel addresses' subnets, e.g. 172.16.40.2 fe80::2001:db8:1:2 and one or more addresses outside of the instance's tunnel addresses' subnets, eg. 44.31.16.1 2620:12a:2032:4::1

Routes that would be created to route traffic for 44.31.16.1 and 2620:12a:2032:4::1 to the resulting Wireguard interface, e.g. wg1, are not created.

**Relevant log files**

The following error lines observed in /var/log/wireguard/latest.log:

```
<35>1 2024-03-06T13:48:25-06:00 opnsense.example.com wireguard 9297 - [meta sequenceId="2"] /usr/local/opnsense/scripts/Wireguard/wg-service-control.php: The command '/sbin/route -q -n add -'inet' '/' -interface 'wg1'' returned exit code '68', the output was 'route: bad address:'
<35>1 2024-03-06T13:48:25-06:00 opnsense.example.com wireguard 9297 - [meta sequenceId="3"] /usr/local/opnsense/scripts/Wireguard/wg-service-control.php: The command '/sbin/route -q -n add -'inet6' '/' -interface 'wg1'' returned exit code '71', the output was 'route: : Name does not resolve'
```

**Environment**

OPNSense 24.1.2 and Above (amd64)

**Comment by fichtner** [comment]
"172.16.40.2" isn't a valid allowed IP, so what are you actually inputting there?

**Comment by fichtner** [comment]
Oh it's valid for whatever reason but it should have a subnet mask...

**Comment by fichtner** [comment]
If you can confirm these work with /32 attached we can consider a migration for these single-IP entries, but OTOH if you have a tunnel address 172.16.40.1/24 all of its addresses are routed automatically by the kernel.

**Comment by rfrederick** [comment]
> Oh it's valid for whatever reason but it should have a subnet mask...even the help text says so.

You're right, and when the entries were made way back when they did not include a /32 or /128 for the subnet mask. Explicitly adding subnet masks caused the routes to be added.

Perhaps we can tweak the input checking for the Allowed IPs to ensure that a subnet mask is included?

**Comment by fichtner** [comment]
Yep, but we have to migrate the "wrong" entries as well before tightening the validation for it.

**Comment by fichtner** [comment]
I'm still curious why you need a 172.16.40.0/24 and a 172.16.40.2/32 to make things work. Both should be in the routing table?

**Comment by fichtner** [comment]
Ok maybe I got hung up on things... you need the allowed IPs for the outside addresses 44.31.16.1, inside addresses should be fine without "allowing" them. It's quite counter intuitive in the official WireGuard wording.

**Comment by rfrederick** [comment]
> I'm still curious why you need a 172.16.40.0/24 and a 172.16.40.2/32 to make things work. Both should be in the routing table?

You are right, the /32 is redundant in this case.

**Comment by fichtner** [comment]
Ok, thanks for confirming. Let me work on a migration for this tomorrow to get this ironed out rather sooner than later.

Cheers,
Franco

**Comment by fichtner** [comment]
The same is true for system routes... but to be very honest unless there is a real issue we don't need to do this.
